### PR TITLE
[NAE-1807] Reference in Arc is not cloned properly

### PR DESCRIPTION
**Pull request by renczesstefan**
# Description

Implemented clone methods on all petri net objects.

Fixes [NAE-1807]

## Dependencies

No new dependencies were introduced

### Third party dependencies

No new dependencies were introduced

### Blocking Pull requests

No new dependencies were introduced

## How Has Been This Tested?

This was tested manually.

### Test Configuration

| Name                | Tested on |
|---------------------| --------- |
| OS                  |   macOS Ventura 13.0     |
| Runtime             |  Java 11        |
| Dependency Manager  |  Maven 3.8.4         |
| Framework version   |  Spring Boot 2.6.2        |
| Run parameters      |           |
| Other configuration |           |

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] My changes have been checked, personally or remotely, with @mazarijuraj 
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have resolved all conflicts with the target branch of the PR
- [x] I have updated and synced my code with the target branch
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes:
    - [x] Lint test
    - [x] Unit tests
    - [x] Integration tests
- [x] I have checked my contribution with code analysis tools:
    - [x] [SonarCloud](https://sonarcloud.io/project/overview?id=netgrif_application-engine)
    - [x] [Snyk](https://app.snyk.io/org/netgrif/project/ea82b3b8-658d-41f5-89a7-76cd600da01f)
- [x] I have made corresponding changes to the documentation:
    - [x] Developer documentation
    - [x] User Guides
    - [x] Migration Guides

[NAE-1807]: https://netgrif.atlassian.net/browse/NAE-1807?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/com/netgrif/application/engine/petrinet/domain/Component.java [hunk 1]
```diff
@@ -40,4 +41,13 @@ public Component(String name, Map<String, String> properties, List<Icon> optionI
         this.properties = properties;
         this.optionIcons = optionIcons;
     }
+
+    @Override
+    public Component clone() {
+        Component clone = new Component();
+        clone.setName(this.name);
+        clone.setProperties(new HashMap<>(this.properties));
+        clone.setOptionIcons(this.optionIcons.stream().map(Icon::clone).collect(Collectors.toList()));
```
Comment (mazarijuraj) [inline src/main/java/com/netgrif/application/engine/petrinet/domain/Component.java:50]: Can optionIcons be null? Would be better to check for nullpointer

File: src/main/java/com/netgrif/application/engine/petrinet/domain/arcs/reference/Referencable.java [hunk 2]
```diff
@@ -3,4 +3,6 @@
 public interface Referencable {
 
     int getMultiplicity();
+
+    Referencable clone();
```
Comment (mazarijuraj) [inline src/main/java/com/netgrif/application/engine/petrinet/domain/arcs/reference/Referencable.java:7]: Why was this added? Cloning is not the concern of Referencable

Comment (renczesstefan) [reply src/main/java/com/netgrif/application/engine/petrinet/domain/arcs/reference/Referencable.java:7]: Because the Arc class contains only Referencable interface, and not the regarded implementations (Place, Datafield)

File: src/main/java/com/netgrif/application/engine/petrinet/domain/Component.java (hunk 1 above)
Comment (renczesstefan) [reply src/main/java/com/netgrif/application/engine/petrinet/domain/Component.java]: Corrected

File: src/main/java/com/netgrif/application/engine/petrinet/domain/arcs/reference/Reference.java [hunk 3]
```diff
@@ -30,4 +30,11 @@ public int getMultiplicity() {
         }
         return multiplicity;
     }
+
+    public Reference clone() {
+        Reference clone = new Reference(this.reference, this.type);
+        clone.setReferencable(this.referencable == null ? null : this.referencable.clone());
```
Comment (mazarijuraj) [inline src/main/java/com/netgrif/application/engine/petrinet/domain/arcs/reference/Reference.java:36]: Does it work as intended? If you clone the referencable you create a new object and it does no longer reference an object from the cloned Petri net.

**Review by mazarijuraj — CHANGES_REQUESTED** [review]
I think `Referencable` does not need to be cloned because it is marked as `@Transient` and only references an object from a Petri net, cloning it creates a new object which is not part of the original nor cloned Petri net. I think it works because it is always set properly by calling `initializeArcs(Map<String, DataField> dataSet)` in the WorkflowService.

File: src/main/java/com/netgrif/application/engine/petrinet/domain/arcs/reference/Reference.java (hunk 3 above)
Comment (renczesstefan) [reply src/main/java/com/netgrif/application/engine/petrinet/domain/arcs/reference/Reference.java:36]: Ok, now I understand, I will remove this method and callings for it.

Comment (renczesstefan) [reply src/main/java/com/netgrif/application/engine/petrinet/domain/arcs/reference/Reference.java]: Corrected.

**Review by mazarijuraj — APPROVED** [review]

File: src/main/java/com/netgrif/application/engine/petrinet/domain/Component.java [hunk 4]
```diff
@@ -40,4 +41,13 @@ public Component(String name, Map<String, String> properties, List<Icon> optionI
         this.properties = properties;
         this.optionIcons = optionIcons;
     }
+
+    @Override
+    public Component clone() {
+        Component clone = new Component();
```
Comment (machacjozef) [inline src/main/java/com/netgrif/application/engine/petrinet/domain/Component.java:47]: please re-do via the constructor:
 https://www.artima.com/articles/josh-bloch-on-design#part13

**Review by machacjozef — CHANGES_REQUESTED** [review]

File: src/main/java/com/netgrif/application/engine/petrinet/domain/Component.java (hunk 4 above)
Comment (renczesstefan) [reply src/main/java/com/netgrif/application/engine/petrinet/domain/Component.java]: done

File: src/main/java/com/netgrif/application/engine/workflow/domain/DataField.java [hunk 5]
```diff
@@ -206,4 +207,19 @@ public int getMultiplicity() {
             throw new IllegalArgumentException("Variable arc must be an non negative integer");
         }
     }
+
+    @Override
+    public Referencable clone() {
+        DataField clone = new DataField();
+        clone.setBehavior(this.behavior);
```
Comment (mazarijuraj) [inline src/main/java/com/netgrif/application/engine/workflow/domain/DataField.java:214]: Is `DataField` ever cloned? I think `DataField` is only used in `Case` and does not need cloning. If cloning is needed then this function will create a shallow copy of its collections.

**Review by mazarijuraj — APPROVED** [review]
Please see comment on `DataField`
